### PR TITLE
Feature ID changes from EXT_mesh_features spec updates

### PR DESCRIPTION
**Pull request by ptrgags**
This PR targets a staging branch, `feature-id-changes`

the `EXT_mesh_features` extension is under some revisions, being split into `EXT_mesh_features` and `EXT_structural_metadata`.

This PR doesn't cover everything, but it's a start. It includes:

* the `FEATURE_ID_n -> _FEATURE_ID_n` attribute semantic
* added `nullFeatureId` and `featureCount` to `ModelComponents.FeatureIdXxx` (though these are not yet used for anything)
* feature ID textures now have the texture info moved into a `texture` property.
* Parse feature IDs from a multi-channel texture

[Sandcastle for multi-channel Feature IDs](http://localhost:8080/Apps/Sandcastle/index.html#c=3RuLcts28ldQ39yUTmTqaceV48w5fqS5sRpPbCc3F3d8EAlJqCGSB5C21I7//XYBkAQl2lLi9Nq0k8oSCO4u9r0L4JApnk3941nCJJ+yKKXihNE0k0z5LKJDwQZxyIT7nOyTVGZs7yoK4kil5JazOyZhNGJ35NDA+6DHvKuNQP8+jOFFHjF5tdEgv11FhMQyZPJtFLKEwUeUXkgaKZEFLArmfTKiQrHGVXS/CViuIoPBD0Qc3PhBJiW+ANQATovvn5ngNDqiKfNHMp6+VfHuTqvtIaarjU6r095qt7daP1x0Wv3Oi373hd9q77R2er0X293e7s5u74fui39fbVxFBl+zSfSqSRIrnvI44tGY7H/efwjH8CcHUlJ7SGUK32jU1dQesbFkTGlqt9qdrt960evttH9o4ECv57e2W90XrR39s2VoNJAnyQLbf2Q0BFrPeBpM3sdCeK0GwX/lKyM+Y+GJpFOmOT6K5bSkqxhSPrAaNEHPi0+Kd94wkCFNY2k5G8UynYBMza87ptIKEw8zlcZTcj6hIGxyxEY80pxQ5HO4CICOJB2PUQjphJFpnClG7rgQhM0SGoVNNZE8urEPQW5+vlrzXMvSElFhl6HPPPG0WmYR1+vvGy0lBJB/YAGsmKCkiAANA7gh0ENAV1OAmMYlVead7BqfFyAISecJ6+dILw2KCxjzPxwfdhr5rFsqMpjm0pfrScdr+ShHv7Vpp9/rv+bzlsG0mVnGBZulffKfgvi3I4dnXGnKxyzMqZZ8PEkbJefIWMZ3at23BRtVXjZysK/fxjy0pA3A9r0P+uvbKMnAZSj9t0F4FMPP4NfptYZg5rzLUjPJfNk08ApukuKJn5vW4JA830f2tMkzy35/Bl8tHp+mqeTDDJjpg8pOqRgc7lk+mj//KZ2N0RwgA52gFdaXKw8TNFHIMXRXPCKKAfxQEZhIaMSnFMnP1QYnraM2J6fvDi4W9aZV0QyNHKQmt0I0O6TArCjHpYAjCVProDs/GJydHr+/7hw9oqsXFYZ5jrQyKfrgHXy/Cf/O6TQRDBw1bZrQkBPiJ9HYehK9ijpFhyUdJImYa5Wz6yntD/SnQUDI8PWWVaYEMYQbHqHtWjCUDHlKYJ40klEx4en3yoqEhT7OG4EaYcirsSyt3Plzrd4n9odR8NEDCj4A4JJDDJ3aLzXKzYIOUn6IRIPKjZZ1OH963SLPQetR6fEtz2iQ8RN7JcQclx/y0QgteT/nTOfIKzShUSDd9OV4+KB9APem9IYBC4MJC26YHMZUFuoFfjmdABsJFcmEghRoCjwIJKOKKfMQJKOhhJyO4wh4EWXTIaQG0SiLAh0oEf6hA9zqlmdxnPNf2aZhmTHWCiHGSC95lO4eSEnnXg/Y47xZ/WXWlAOa0tlRTtU+6cBUFyfZIm3DV7ReTzBYGkxr7cGfly5QGHj+fDMXajH5FzP5l8XJvziTc0o4ZEUzmM8XiH9OfikkixEWV54HxoRKns5JbJz2AnvJmN+CBKpScyAloF9MRn6VjLDkBq/g5iPiFQ//DqyCYA15hqvIAFTBqpEW9d+Mgm5IMCzykZE4AhuOmAkk+Rx4CLTRKGLiuxKIS+0nlKRmzM8one3tgpp7ZyExAJN3HFME9n24SAIgHAoKaSR5DXZJUUGAKfCACtDRcO4A0skKFcB1TWdLoZeIADzyG6mHcUDF/KvIeQuljfiM/tsFNUCzklqpAOBPENlhMT8XnK9dMnC/i8v2YCqM5XA2SdPVWtdq9YdkoDrRakeNzj/U9tKvEGC98B0P00nf1UT7YMIwg6h7gmlbHH2PqgyKlcSYOjVAJuSORmCykqukargsHDObO0yB9yMe6OB4wgW8318gf7A0w//p+OD98XkeF6d0vBLG8pQKEO3w7vOMIFh2SCCOh1zVbuEsL5UJRpW1OspOKFrllKqbOkwDGP/C7CO7dgE9LdDXLL4mPtcGTKgM/lQhc6/yXm8xeLiRsWIIZXh8PLRO+cxbHG4gqq6n8/cKQl9uPhZoD2MB7kTw5XCr8xwMwzrPwboNvD8WzyAxOdeBVkNwfU6dfh2gm/pmFAwWFDHIG5Ve44RhPp2m6I8Ljmu/i3O5KnsK6KBkxhrfSFb3e6toVQOLXK8y28Svhbn0MWWtdXQQ3wJg1SQWkHxA/VOk6nXK+CPO+ra93V9YkTDlq2oOeaVLDZc8/C/kKoDnzqv3j6jNxYQrwma6LsQeA2hMyKaoGxKYYDQqjm6Bd7pzlGE1jUlWLDmOQdWY+zm7HJ88n1lARYsDKpKQPJ/jsO7jPHs2jFPsTeGkOHn2zC9pMa8ONVrFVQpYTOHykQ3fnLrE6JeWF0GFqqxBgfrfIUy0PkrQDhm2MEnCZxD/USFp2tBwtIUAqIBi0Hj/5rW/UBa9kTTEded5RqUOUlghYGa8s1cOlpndcmnUhUxSmZpI2WKovrhRa1U1aq1yRtXUMZgPlWWi9RMgxpBLZhbvueLcdI3bLO5T18mUW5gpVxC8kQyixTKKuYtCq0ZFMVYhaiMi7iIaggdBFWqZngJ+jac8zXsKZW7+pdl5+b2am6vlpNwZ0qp2ojWtcIZn5ZgPqlaT846r2mbz3SUdrKYrbueFihiCM08VufzgtmH8RQxf7PVzAE/z+AsL/dN4+xWtm5zsxuNhYGVD56H+OpDJR6bddvb6vWGWykVnHp4N5V+4tf6I1FHfBQ9udGjRVAO5OvleZNc6OnL75ToyEjE1fDuCaEWjALVEsGicTjzDLzeKm9lAJ6d2U0pN4zidQJxLDCu6Lf3HBZgDgDWXb9IwxIYpGcciBN8KoTjNJaYSFmSCSjIBRyTQGakanS5mLRRN+XhRNe2ifLbxow31U0FBLVXooJRRm5x1YxErxbHDuESCjLPxJIKCAjNdAeHba/stslXC05rRbpB22VKtogMfio2hKbMxzHIgt9cAXeIj9vx8P1+kXp/VRGzo5ihWbFQkMeA8FHEWfqS37NvbqFixhfU77SLlKZue/70iM23FczdE6a6gTROgkBwz3ahr/9yAwpMHkwKOyQfRCwtIJuJsiJUOrQl4lQw+uwUh1WxPlTta/mxeKtwgL/W1uEFVQd4gxjBDr4lUUkCGyeeM3FG9+2GNXKc6BRjIbSZgnBRcBK4XxegDI+AXvMvTLNSpy5CiIsRm6ZBmQ2SguLlHE1ZAsk1mlxxP8yVvpCIgisTMicrkCFBuGlRIXwEGU12gG5aOPTFYDL8tdnRsEIhH7v4O1PwBM1sIFkSab1pTUyDoYGJERbZekU9b7eKrHlx0heXSIXMG43+2Qiy/4jaA31pyqZrv+w44zHQjz4BE3Uww+c1u/RkAMKPGkDbtTPtr79F9TkCPG53beuMHG8GI9iE9USTJhDL6AeJGIxnOdd8xD/Zc6iSxFiUAODcFRRt8EiDc1kQvk+q4pj8iT6pZdsCliSwcXCRoIyqR2QBYCptLIXMsrs+0T9UF9hayebEKLl56pZ+uXf/mlFJkYqaQHTo+kIRCJZVas8Xty6JAGesTF6KsUhYtEOVvtj1Rl53iEqJoUXPo3JyNRjxAP6XIHcNd0UmsIG5TOeRgQ5KLeZ2PGq3nox5IVls+bqLVwNDrvm491H1aDo8djIxd/Ohtgi7WjdYanJ7YxcDawY+2ia7WDLuuGa4ItNNMpNyejXobqm810upsDKxfH4opQJ26o/7lT6dvL1ZZM6Z2CSY8QHjsOUZgi9p2yc/fvUyyZ5u0RV8c/+vi8v3x9eG7y58uQD4vany1raHeYhMMcBQb55V3XatfBMDDvAiutti6xqAX/EWJ7qVRwIrDKGGhuSF4LzeYUaFu5Vco50gTXHJvd/vFdmVtxHoTazNtm00unQe4BxXEqLBAWefJlLXXpKxlLXFdyrpPpqyDlHW2V1OVk7YGVb0nU9VFqna2t7ur5fgZ3Np+Ml29tegqpbgev3Y0XeQphG0/rGDNZgUYWvXjoHDTXZO0UjHaKxTjtyodpoOrJnEmQmwjU12/8AArUUvgU3iwsx7Fm9X049Eo7ZDwrOLA6ptFbBqrzz0O++DpWO3isfzWXBxSIeI4qj0vZmY2TcR6bSZWf/ljMbRnyEIZR2w1mCOc5n53QCT0LlLrgjjDyc7LUy5uLmQW3PQNC1dBGeTzF387MMcyhpLvA4MyS7CVcN+4s6u/3EVW2gcrgZ5Vpi/8dMBC7j1mZbJUwG2eJyxQTQ1yYOkU6eg0xvSieVp5y/Zim2NxcVL/CBCmI43xXisnbkfgwRyF/Uy0JjxGXhx9Do3ekk9ITN5Y1xnN1cYAk7ut/ESGxUPeHhGLSRVHEwEUE5CK90m+L+PmPuaZXplnWOYvsKJRn0iW5lq3gHLfoFFDvD1mbUl9hFKygtSKgjVqz8CuQWfjQULPZBywMMN65unEWlfReOC4zJMIdU4OFP5SfW1anaMXTyIW993zDRqqCLLga5PqbMx/LQUoIs5X1ITqZtKTKB2YnY+yjU9uOSUDvTWCkL6elS1so9RTbS/crCD6WF93sHdXvhq9RShrLN+neAq1B/asNdFRhOgw8gQyq5GsUdsY/xyV+FmHlAK7izGTAuzDKfYtZfaekgK/wfxE8inHfqbyJcOezoEQ3qazMW/uTOw/+BYNQ88QZYv0pZtY+vYQxs/yuD1GxD4euS96AS6d/cqvYkpeZks+69fcBZos3CpybwR5ZWKZt4Ua5dAkkc4vC/lYCJ5g88v/+OZ8t+dMqLmclD8sNud08Z+f2ja3fXS/GVzLlCvmpxMWeaXO6BmF4uQXyQCDpP5IzC/i12iMsLrzBM/ymvn+sDLYKPUOfRcC7usNqZykPXfn/BwsJKAqhTQLJHgRx2JI5YBFmacTkCKNPsZTJAQyjlAw+YX5dJlGq5TK/E4cKIjuh0cB86P4TitdVclilV4mId6WAxI1Iaf6oAuT3pK52ZtUpmmFGM5tv2qfeEuYyFZJianVWqZMqcslfMXS/KDD1YbpweAFwWVchsN1Fv25MGqTr88DYishTDZxj/Qg0JsWS9kmPDo0Oxj7D2w8Vy43qgBPqJzj3owWyI9GMZA0XQwemFzT1exbhm7Aigmr7e8Kr1ZovGnG7TlnTuy2IYfgHuaJ7qITgocF/KLdaxio8di12KtFXgXY5iO4K1icFR9qg8RbohL8C8wwV0/1wpXDWnsGND/Tj2x1zguYus+RSBGHltaCd0kj5pVC0oJt5EI6XxCGbrGeHp9cXB+9+/iT2xWGRVBwizaSPCzn/7eU9TZFwYtNp/08tswzxzyKy6mP774Zug2MUr9rb7SqbJhKGqReGV0s81kUni3GiFICxZDDUjOUe3tNvvFbRWPdjC/lJYsWHZoUiLinIRYSr9Wv3K9UkcG7y/Pj68G7D8d/jG1/qX1ZQ1kyoNzu1rGMyzOz5o3GxkuVzgV7lQv0H3yaxDLFrMTz/WbKpok+7NMcQlLJUj9QqkzM/paaiOl21YY0uDFZc5/I8ZB6vU6D5P+3/F2315VARIPI3Se9ZOYMD/WF8y0sEDJVfXi/hJrr/YjKoVvTwNuigo8h+E95CAJcxrqVxkmfdCqY80fmXGH1qUX9suky7GXIbwkP92suz+MZFaXgySgTAjdnrzZevWzC/KVXRaxztndAuaBznDZpvzo1g77vv2zCz/o3LRMWIP8P) -- some of the feature ID textures are a bit contrived, I need to find a better way to check them for correctness. I also need to try picking/styling with the one texture that has null feature IDs.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @ptrgags!
* :heavy_check_mark: Signed CLA found.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

File: Source/Scene/ModelExperimental/FeatureIdPipelineStage.js [hunk 1]
```diff
@@ -318,6 +318,13 @@ function processImplicitRange(
   );
 }
 
+const shiftsByChannelCount = [
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/FeatureIdPipelineStage.js:321]: I'm wondering if this is best done as a builtin function, `czm_unpackUint()` overloaded for all 4 types, right now this makes for a messy one-liner like this:

```
" featureIds.featureId_0 = dot(floor(texture2D(u_featureIdTexture_0, v_texCoord_0).rgba * 255.0 + 0.5), vec4(1.0, 256.0, 65536.0, 16777216.0));"
```

File: Source/Scene/ModelExperimental/FeatureIdPipelineStage.js [hunk 2]
```diff
@@ -318,6 +318,13 @@ function processImplicitRange(
...
+  "1.0",
+  "vec2(1.0, 256.0)",
+  "vec3(1.0, 256.0, 65536.0)",
+  "vec4(1.0, 256.0, 65536.0, 16777216.0)",
+];
+
 function processTexture(
   renderResources,
   featureIdTexture,
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/FeatureIdPipelineStage.js:330]: In the fold below:

I realized one flaw in `processTexture()`:

right now I'm labeling the texture uniforms as `u_featureIdTexture_{featureIdSetIndex}` -- The problem is this is wasteful if two different feature ID sets share the same texture. Now you've used up 2 texture slots when only 1 is strictly needed.

I think it's better to use the glTF texture ID instead. Though I'll need to update `GltfLoader` and `ModelComponents.TextureReader` to store the index, right now it only stores the Texture object. I might do this in a future PR.

File: Source/Scene/ModelExperimental/FeatureIdPipelineStage.js (hunk 1 above)
Comment (ptrgags) [reply Source/Scene/ModelExperimental/FeatureIdPipelineStage.js:321]: Talking with @lilleyse, we decided to create such a function, 

`int czm_unpackUint(genType vec)` -- and just explain the limitations.

Comment (ptrgags) [reply Source/Scene/ModelExperimental/FeatureIdPipelineStage.js:321]: Also note that the `floor(x * 255 + 0.5)` should be a part of this function.

File: Source/Scene/ModelExperimental/FeatureIdPipelineStage.js [hunk 3]
```diff
@@ -318,6 +318,13 @@ function processImplicitRange(
   );
 }
 
+const shiftsByChannelCount = [
+  "1.0",
+  "vec2(1.0, 256.0)",
+  "vec3(1.0, 256.0, 65536.0)",
+  "vec4(1.0, 256.0, 65536.0, 16777216.0)",
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply Source/Scene/ModelExperimental/FeatureIdPipelineStage.js:325]: We decided not to impose any limitations, just document them.

File: Specs/Scene/GltfLoaderSpec.js [hunk 4]
```diff
@@ -868,6 +868,8 @@ describe(
         expect(featureIdTexture).toBeInstanceOf(
```
Comment (ptrgags) [inline Specs/Scene/GltfLoaderSpec.js:868]: Add a unit test for the large feature ID texture, so we have an example that includes a null feature ID

File: Source/Scene/GltfLoader.js [hunk 5]
```diff
@@ -813,15 +813,22 @@ function loadMaterial(loader, gltf, gltfMaterial, supportedImageFormats) {
 // for EXT_mesh_features
 function loadFeatureIdAttribute(featureIds, propertyTableId) {
   const featureIdAttribute = new FeatureIdAttribute();
+  featureIdAttribute.featureCount = featureIds.featureCount;
+  featureIdAttribute.nullFeatureId = featureIds.nullFeatureId;
   featureIdAttribute.propertyTableId = propertyTableId;
```
Comment (ptrgags) [inline Source/Scene/GltfLoader.js:818]: the property table ID should come from `featureIds.propertyTable` I think

File: Source/Scene/ModelExperimental/FeatureIdPipelineStage.js (hunk 2 above)
Comment (ptrgags) [reply Source/Scene/ModelExperimental/FeatureIdPipelineStage.js:330]: opened https://github.com/CesiumGS/cesium/issues/10090 for this

File: Source/Scene/ModelExperimental/FeatureIdPipelineStage.js [hunk 6]
```diff
@@ -318,6 +318,13 @@ function processImplicitRange(
   );
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/FeatureIdPipelineStage.js:318]: also `SelectedFeatureIdPipelineStage`'s shader should be updated to check for `nullFeatureId`, treating it like the current behavior when the feature ID is greater than the number of features in the property table

**Comment by ptrgags** [comment]
@lilleyse the code is now ready for review. Note that I want to make two Sandcastles for testing:

- (TBD) Update the sandcastle from the description to have better visualizations for the larger feature ID values
- (TBD) Make a sandcastle for picking the 7th feature ID texture to check if the null feature ID code works correctly

**Comment by ptrgags** [comment]
image
in that first sandcastle, it seems that fractal pattern I found yesterday happens only in the case where I use all 4 channels (RGBA for the first, AGBB for the second). Somehow the gradients in each channel + precision limitations combine to make a fractal pattern. 

image

These precision issues were to be expected given that `highp int` is actually represented as a float32 in GLSL 1.00. All the other cases with 1-3 channels work fine.

**Review by lilleyse — COMMENTED** [review]
Looks good! Just a couple comments. I tested picking in the 3D Tiles Next sandcastles as well as pnts, i3dm, and b3dm with `enableModelExperimental: true`.

File: Apps/SampleData/models/BoxInstanced/BoxInstanced.gltf [hunk 7]
```diff
@@ -86,20 +86,18 @@
...
           }
         },
         "EXT_mesh_features": {
-          "propertyTables": [
-            0,
-            1
-          ],
           "featureIds": [
             {
+              "propertyTable": 0,
```
[the remark this answers is not in the document]
Comment (ptrgags) [reply Apps/SampleData/models/BoxInstanced/BoxInstanced.gltf:95]: ah I'll update the documentation and glTF models.

the `array -> dictionary`... that's best done in its own PR since that's going to mean we have to swap out `featureIdIndex` (Number) for... idk `featureIdSet` (String)

And the removal of implicit feature IDs is best done when I split out `EXT_structural_metadata` so I can replace it with property mappings. (which might get messy given that things are now accessors instead of bufferViews)

I'll open some issues for these.

[the remark this answers is not in the document]
Comment (ptrgags) [reply Apps/SampleData/models/BoxInstanced/BoxInstanced.gltf:95]: Opened https://github.com/CesiumGS/cesium/issues/10095 for the first one

[the remark this answers is not in the document]
Comment (ptrgags) [reply Apps/SampleData/models/BoxInstanced/BoxInstanced.gltf:95]: Opened https://github.com/CesiumGS/cesium/issues/10096 for discussing the second one (as there may be some conceptual overlap between this and the implicit tiling part)

File: CHANGES.md [hunk 8]
```diff
@@ -11,6 +11,7 @@
 - Added support for shadows in `ModelExperimental`. [#10077](https://github.com/CesiumGS/cesium/pull/10077)
 - Added the ability to load implicit tiling subtree files from JSON. [#10086](https://github.com/CesiumGS/cesium/pull/10086)
 - Added support for subtree metadata in tilesets with implicit tiling. [#10093](https://github.com/CesiumGS/cesium/pull/10093)
+- Updated Feature ID parsing to match changes to the EXT_mesh_features specification. [#10087](https://github.com/CesiumGS/cesium/pull/10087)
```
Comment (ptrgags) [inline CHANGES.md:14]: @lilleyse just a note, I do plan to clean up this changelog before we merge the staging branch. it'll probably be best to make a list of sub-bullets (like we did when we first introduced 3D Tiles Next) but that'll be easier once everything is in place.

**Comment by ptrgags** [comment]
@lilleyse updated!
